qiskit-superstaq/qiskit_superstaq/superstaq_job.py

```diff
@@ -90,35 +98,85 @@ def result(self, timeout: Optional[float] = None, wait: float = 5) -> qiskit.res
             }
         )
 
-    def status(self) -> qiskit.providers.jobstatus.JobStatus:
-        """Query for the job status."""
+    def _check_if_stopped(self) -> None:
+        """Verifies that the job status is not in a canceled state and raises an
+        exception if it is.
 
-        job_id_list = self._job_id.split(",")  # separate aggregated job ids
+        Raises:
+            SuperstaQUnsuccessfulJob: If the job been canceled or an error has occured.
+            SuperstaQException: If unable to get the status of the job from the API.
+        """
+        if self._overall_status == "Canceled":
+            raise gss.superstaq_exceptions.SuperstaQUnsuccessfulJobException(
+                self._job_id, self._overall_status
+            )
+
+    def _refresh_job(self) -> None:
+        """Queries the server for the current job status"""
 
-        status = "Done"
+        # When we have multiple jobs, we will take the "worst status" among the jobs
+        # For example, if object of the jobs are still queued/canceled, we report Queued/Cancelled
+        # as the status for the entire batch.
+        # The overall status will only be reported as Done if all jobs are done
+        # following the "worst status" rule.
+
+        job_id_list = self._job_id.split(",")  # separate aggregated job ids
 
-        # when we have multiple jobs, we will take the "worst status" among the jobs
-        # For example, if any of the jobs are still queued, we report Queued as the status
-        # for the entire batch.
         for job_id in job_id_list:
-            result = self._backend._provider._client.get_job(job_id)
-            temp_status = result["status"]
 
+            if job_id in self._job_info and self._job_info[job_id] not in self.TERMINAL_STATES:
+                result = self._backend._provider._client.get_job(job_id)
+                self._job_info[job_id] = result
+
+            if job_id not in self._job_info:
+                result = self._backend._provider._client.get_job(job_id)
+                self._job_info[job_id] = result
+
+            temp_status = self._job_info[job_id]["status"]
             if temp_status == "Queued":
```

Comment:
Yeah, I guess let's set a guideline to follow -- like with the scenario of 1 job out of ten being canceled or failed, what do we set as the status for the whole job (relevant b/c `_wait_for_final_state` checks `status`). The `status()` in [`qiskit_ibm_provider.job.CompositeJob`](https://github.com/Qiskit/qiskit-ibm-provider/blob/main/qiskit_ibm_provider/job/ibm_composite_job.py) returns the status of a job containing multiple subjobs in the following order: 

                * INITIALIZING - if any sub-job is being initialized.
                * VALIDATING - if any sub-job is being validated.
                * QUEUED - if any sub-job is queued.
                * RUNNING - if any sub-job is still running.
                * ERROR - if any sub-job incurred an error.
                * CANCELLED - if any sub-job is cancelled.
                * DONE - if all sub-jobs finished.
 Do we store information on all the job results (even with 1 or more failed job_ids in the middle of a `job_id_list`)? Then only retrieve the successful bunch when the user calls `result()` and indicate specifically the job_ids that have failed?

Reply:
i think following that order is a great idea

if we want successful jobs to be available even if others fail, it seems like this means we can only break the loop if a sub-job's status is INITIALIZING (otherwise one of the later sub-jobs might change the overall status). Otherwise we could arguably also break the loop after a job returns ERROR because there's no reason to wait for the others, but it doesn't seem like we gain much by doing this..

> Then only retrieve the successful bunch when the user calls `result()` and indicate specifically the job_ids that have failed?

we should just need to set `success=False` when we append the results of failed jobs to `results_list` in `superstaq_job.result()`, e.g.:
```python
           results_list.append(
                {
                    "success": result["status"] == "Done",
                    "status: result["status"],
                    "shots": result["shots"],
                    "data": {},
                }
            )
```
then calling `job.result().get_counts()` will still raise an error if any sub-job failed, but e.g. `job.result().get_counts(1)` will return counts for sub-job 1 if it was successful

Reply:
I've implemented status checking following that order and added more tests to check for different status ordering as you suggested previously. 

I also added "success" to the `results_list`. But maybe we should also mirror the `partial` parameter used in [qiskit_ibm_provider.job.CompositeJob](https://github.com/Qiskit/qiskit-ibm-provider/blob/main/qiskit_ibm_provider/job/ibm_composite_job.py) someway to retrieve successful sub-jobs for an overall failed job as an easier alternative to trying `job.result().get_counts(1)`.

Reply:
ah good idea. would this override the `check_if_stopped` call in `_wait_for_results`?

Reply:
So it seems [`qiskit_ibm_provider.job.CompositeJob`](https://github.com/Qiskit/qiskit-ibm-provider/blob/main/qiskit_ibm_provider/job/ibm_composite_job.py) first checks `if self._status == JobStatus.DONE or partial:` and then only raises an error from `JobStatus.CANCELLED` if that initial if-statement was not satisfied. So, for our case, I am imagining passing a method param and having variations of the exceptions. Though, if this further functionality is not required for open-beta, I can leave it for another PR. 

Also curious, does `css.job` have this functionality?

Reply:
> So it seems [`qiskit_ibm_provider.job.CompositeJob`](https://github.com/Qiskit/qiskit-ibm-provider/blob/main/qiskit_ibm_provider/job/ibm_composite_job.py) first checks `if self._status == JobStatus.DONE or partial:` and then only raises an error from `JobStatus.CANCELLED` if that initial if-statement was not satisfied. So, for our case, I am imagining passing a method param and having variations of the exceptions. Though, if this further functionality is not required for open-beta, I can leave it for another PR.

i think it's fine to leave the `partial=` flag out of this pr. imo for now we should just remove the `check_if_stopped` from `_wait_for_results`, so that `result()` always returns something once all jobs are in terminal states (even if some failed or were cancelled). this might be a bit different from what qiskit does by default but seems maybe more useful (at least for now)

if/when we do add the `partial=` flag, it seems like we'd also want to make this return immediately even if some jobs are still queued/running.. (but obviously all of this will require further refinement/feedback/etc as it gets more use)

> Also curious, does `css.job` have this functionality?

currently css.job only supports one job at a time - see #352

Reply:
Removed `check_if_stopped` for now. Will create a related issue to implement a `partial` style flag.